### PR TITLE
Use DateTimeOffset instead of DateTime

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**
Switched to using DateTimeOffset instead of DateTime for IoT hub and DPS clients

### CODE REVIEW

**Review by patilsnr — APPROVED** [review]

File: iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs [hunk 1]
```diff
@@ -64,17 +64,17 @@ public abstract class AuthenticationWithTokenRefresh : IAuthenticationMethod
         /// <summary>
         /// Gets a snapshot of the UTC token expiry time.
         /// </summary>
-        public DateTime ExpiresOn { get; private set; }
+        public DateTimeOffset ExpiresOn { get; private set; }
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs:67]:
```suggestion
        public DateTimeOffset ExpiresOnUtc { get; private set; }
```

File: iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs [hunk 2]
```diff
@@ -64,17 +64,17 @@ public abstract class AuthenticationWithTokenRefresh : IAuthenticationMethod
...
 
         /// <summary>
         /// Gets a snapshot of the UTC token refresh time.
         /// </summary>
-        public DateTime RefreshesOn => ExpiresOn.AddSeconds(-_bufferSeconds);
+        public DateTimeOffset RefreshesOn => ExpiresOn.AddSeconds(-_bufferSeconds);
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs:72]:
```suggestion
        public DateTimeOffset RefreshesOnUtc => ExpiresOn.AddSeconds(-_bufferSeconds);
```

File: iothub/service/src/Authentication/Security/SharedAccessSignature.cs [hunk 3]
```diff
@@ -44,7 +44,7 @@ internal sealed class SharedAccessSignature
         /// <summary>
         /// The date and time the SAS expires.
         /// </summary>
-        internal DateTime ExpiresOn { get; private set; }
+        internal DateTimeOffset ExpiresOn { get; private set; }
```
Comment (drwill-ms) [inline iothub/service/src/Authentication/Security/SharedAccessSignature.cs:47]:
```suggestion
        internal DateTimeOffset ExpiresOnUtc { get; private set; }
```

File: iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs [hunk 4]
```diff
@@ -190,7 +190,7 @@ private async void OnFeedbackMessageReceivedAsync(AmqpMessage amqpMessage)
 
                         var feedbackBatch = new FeedbackBatch
                         {
-                            EnqueuedOnUtc = (DateTime)amqpMessage.MessageAnnotations.Map[MessageSystemPropertyNames.EnqueuedOn],
+                            EnqueuedOnUtc = (DateTimeOffset)amqpMessage.MessageAnnotations.Map[MessageSystemPropertyNames.EnqueuedOn],
```
Comment (drwill-ms) [inline iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs:193]: I tried updating this before, and IIRC it failed. This one we may not be able to change.

File: provisioning/service/src/Twin/TwinCollection.cs [hunk 5]
```diff
@@ -174,9 +174,9 @@ public TwinMetadata GetMetadata()
         /// <returns>DateTime instance representing the LastUpdated time for this property.</returns>
         /// <exception cref="NullReferenceException">Thrown when the TwinCollection metadata is null.
         /// An example would be when the twin collection class is created with the default constructor.</exception>
-        public DateTime GetLastUpdatedOnUtc()
+        public DateTimeOffset GetLastUpdatedOnUtc()
         {
-            return (DateTime)_metadata[LastUpdatedName];
+            return (DateTimeOffset)_metadata[LastUpdatedName];
```
Comment (drwill-ms) [inline provisioning/service/src/Twin/TwinCollection.cs:179]: We'll have a merge conflict on these twin changes, but no worries. It'll be easy to resolve.

Comment (drwill-ms) [reply provisioning/service/src/Twin/TwinCollection.cs:179]: Because I already updated it to be DateTimeOffset.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**